### DISCUSSION TITLE
Access current state/context inside actions with predictableActionArguments?

### DISCUSSION THREAD

**Discussion by geekox86** [statelyai/xstate#3585] [Q&A]
Greetings,

I am trying to use the newly recommended setting `predictableActionArguments`. However and as stated in the docs, I loss access to `meta.state` inside actions. What is your alternative to accessing current state or at least current state's context (i.e. prior to any `assign` action changes) inside actions? My use case depends on comparing previous and new values of the context (i.e. change detection) to trigger different logic.

Thank you!

**Comment by Andarist** [comment]
Could you share some code to illustrate your use case?

**Reply by geekox86** [reply]
Maybe I should have said previous state and previous state's context rather than current state/context prior to any changes by assign actions.
Here is a sample with comments in a1, a2, a3 where I need to access the previous state/context:

```
import { actions, createMachine } from "xstate";

createMachine(
  {
    id: "sample",
    strict: true,
    predictableActionArguments: true,

    context: {
      x: 123,
      y: "abc",
      z: false,
    },

    initial: "a",
    states: {
      a: {
        on: {
          E1: {
            target: "b",
            actions: ["a1", "a2"],
          },
        },
      },
      b: {
        entry: "a3",
      },
    },

    invoke: {
      id: "sampleService",
      src: "sampleService",
    },
  },
  {
    actions: {
      a1: actions.assign({
        x: 789,
        y: "xyz",
        z: false
      }),

      a2: actions.send((context) => ({
        type: "CHANGE_DETECTED",
        data: {
          xChanged: false, // how to access previous state/context here?
          yChanged: false, // how to access previous state/context here?
        },
      }), { to: "someService" }),

      a3: (context) => {
        // how to access previous state/context here?
      },
    },
    services: {
      sampleService: () => (sendBack, on) => {
        // some logic
      },
    },
  }
);
```

**Reply by geekox86** [reply]
@Andarist I would really appreciate your advice on this one :)

**Reply by Andarist** [reply]
I don't have a perfect answer for this. You can always assign the previous context to your context, introduce some conditions logic etc. I understand that this might look like a boilerplate that wasn't needed previously but I consider this to be a fair tradeoff for the time being. This kind of change tracking is not a unique problem to XState - you can find yourself in a similar need in a lot of systems and not every system might provide you with old values to make this easier. Sometimes you just need to implement a little bit of glue yourself.

We will be revisiting this and we might come up with some solutions for this in the future though.

**Comment by james-camilleri** [comment]
Hi! Just following up on this - I seem to have run into the same problem and there doesn't seem to be a documented way around it.  When the `predictableActionArguments` flag is enabled, as recommended, the `state` is `undefined` in the action `meta`.

In the scenario I'm dealing with, I need to know whether to use a "buy" or "sell" price from the context depending if the machine is in the "buy" or "sell" state. I suppose I could copy the price into an additional context property when the machine enters/exists the buy and sell state, but it seems like bad design to have additional copies of the same data in the context, especially when the machine is already rather complex.

Is there a technical reason that the data is missing, something that's planned on being fixed, or is this a specification thing for which there exists a cleaner solution?

**Reply by james-camilleri** [reply]
Hey @Andarist, sorry to chase, but could you let me know if this is an issue or by design please, as I need to plan some internal architecture and I'm not sure if I should bank on this being "corrected" or not. Thanks!

**Reply by Andarist** [reply]
The reason for this being (temporarily) removed is technical. At the moment, the value given to you was somewhat unpredictable - see the example codesandbox here: https://codesandbox.io/s/quizzical-shaw-e7deum?file=/src/index.js . And I could come up with even more surprising examples.

For that reason, we've decided to remove this for the time being. This doesn't mean that it has been removed permanently - but it also doesn't mean that it will be brought back in the very same shape. We need to rethink the use cases, and potential footguns. I would expect that we gonna allow you to query the current state value~ in some way.

For the time being you could, for example, use parametrized actions, like this: https://codesandbox.io/s/practical-sanne-gyl58w?file=/src/index.js

I understand that this also introduces some redundancy here but I don't see how to do this right now without any redundancy whatsoever.

**Reply by james-camilleri** [reply]
Thanks for the update! Yes there is a little bit of redundancy - I'm currently keeping track of the states I need by mirroring them in the context, but this occasionally requires some code gymnastics. I'll look into the parametrized actions, thanks for the tip.

Good to know that there is some (vague) plan for extracting information from the state however. I'll keep that in mind and try to design my solution in a way that can be superseded buy the eventual updates. Thank you for the info 😊
